### ISSUE TITLE
reactor extends itself causes StackOverflowError

### ISSUE THREAD

**Issue by housengw** [lf-lang/lingua-franca#907]
minimal example 
```
target C;
reactor A extends A {}
```

output
```
wonghouseng@beefy-desktop-linux:~/Desktop/lingua-franca/test/C$  lfc src/exp.lf
0    [main] ERROR text.validation.CompositeEValidator  - Error executing EValidator
java.lang.StackOverflowError
	at org.eclipse.emf.ecore.util.EcoreEList.resolveProxy(EcoreEList.java:206)
	at org.eclipse.emf.ecore.util.EcoreEList.resolve(EcoreEList.java:161)
	at org.eclipse.emf.ecore.util.EObjectResolvingEList.resolve(EObjectResolvingEList.java:60)
	at org.eclipse.emf.common.util.BasicEList.get(BasicEList.java:348)
	at org.eclipse.emf.common.util.AbstractEList$EIterator.doNext(AbstractEList.java:698)
	at org.eclipse.emf.common.util.AbstractEList$EIterator.next(AbstractEList.java:685)
	at org.lflang.ASTUtils.allParameters(ASTUtils.java:615)
	at org.lflang.ASTUtils.allParameters(ASTUtils.java:616)
...
	at org.lflang.ASTUtils.allParameters(ASTUtils.java:616)
lfc: fatal error: An unexpected error occurred:
java.lang.NullPointerException
	at org.lflang.FileConfig.toIPath(FileConfig.java:707)
	at org.lflang.FileConfig.toPath(FileConfig.java:690)
	at org.lflang.lfc.Main.getValidatedResource(Main.java:336)
	at org.lflang.lfc.Main.runGenerator(Main.java:271)
	at org.lflang.lfc.Main.main(Main.java:209)
```

**Comment by edwardalee** [comment]
This is because InstantiationGraph no longer detects cycles. Perhaps because of porting to Java?

**Comment by housengw** [comment]
ok, seems like `ASTUtils.allParameters` is called before `LFValidator.checkReactor` is called and `InstantiationGraph` is initialized, and since `allParameters` is recursive, it causes a stack overflow before any checks take place.

Below is the function `allParameters`. `definition.getSuperClasses()` will always be the same reactor if the reactor extends itself. 
```
public static List<Parameter> allParameters(Reactor definition) {
        System.out.println("Doing param");
        List<Parameter> result = new ArrayList<>();
        List<ReactorDecl> superClasses = convertToEmptyListIfNull(definition.getSuperClasses());
        for (ReactorDecl base : superClasses) {
            result.addAll(allParameters(toDefinition(base)));
        }
        result.addAll(definition.getParameters());
        return result;
    }
```

**Comment by edwardalee** [comment]
The situation is even worse than this. Not only are circular extensions not detected, but when multiple extensions include duplicates, the result is invalid.  For example, then following reactor `foo` ends up having 5 input ports, two of which have the same name, rather than the correct four inputs.
```
target C;

reactor D {
    input d:int;
}
reactor C extends D {
    input c:int;
}
reactor B extends D {
    input b:int;
}
reactor A extends B, C {
    input a:int;
}

main reactor {
    foo = new A();
}
```
I'm working on a fix.

**Comment by edwardalee** [comment]
These two problems should be solved with PR #962.
